### PR TITLE
Dev/motor drivers

### PR DESCRIPTION
**Pull request by jsduenass**
create a ROS node that handles the dual max 14870  motor driver
- include a custom version of the dual max 14870   library by pololu
- creates launch file intended to be used to run the robot 
- set up ROS parameters to use multiple instances of the same node. Setting the GPIO pins connected to the raspberry.

### CODE REVIEW

**Comment by jsduenass** [comment]
@cychitivav @Canborda 
ready for review!!!

File: Dockerfile [hunk 1]
```diff
@@ -11,6 +11,7 @@ RUN apt-get update && \
 
 RUN source /opt/ros/noetic/setup.bash
 
+
```
Comment (cychitivav) [inline Dockerfile]: I do not see necessary this space
```suggestion
```

File: LICENSE
Comment (cychitivav) [inline LICENSE:1]: I prefer to use this library like a submodule or create a new

File: README.md
Comment (cychitivav) [inline README.md:1]: We are not using submodules yet

File: config/drivers_pinout.yaml
Comment (cychitivav) [inline config/drivers_pinout.yaml:1]: why use one file with the parameters for the two drivers and then two other files, one for each driver?

File: docs/source/imgs/raspberry-pi-3-PWM-pins.png
Comment (cychitivav) [inline docs/source/imgs/raspberry-pi-3-PWM-pins.png:1]: I prefer to do not upload images.

File: lib/max14870-driver/LICENSE.txt
Comment (cychitivav) [inline lib/max14870-driver/LICENSE.txt:1]: License file duplicated

File: docs/source/index.md [hunk 2]
```diff
@@ -1,2 +1,26 @@
 # Welcome to PAI's documentation!
 This file is a test 2
+
+
+## Raspbery pin out 
+![rpi pin out](./imgs/raspberry-pi-3-PWM-pins.png)
```
Comment (cychitivav) [inline docs/source/index.md]: I would like to use this image
image
```suggestion
![rpi pin out](https://github.com/cychitivav/pai/assets/30636259/018453a2-3278-494c-a706-0de937c5933a)
```

File: launch/run_robot.launch [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0"?>
+<launch>
+  
+     <rosparam file="$(find pai)/config/drivers.yaml" command="load" />
```
Comment (cychitivav) [inline launch/run_robot.launch]: This cannot be loaded, and I think it is not necessary.
```suggestion
```

File: launch/run_robot.launch [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    <!-- Left differential driver  -->
+    <arg name="driver_L"  command="load"   file="$(find pai)/config/driver_R_pinout.yaml" />
+    <node   name="left_driver" 
+            pkg="pai" output="screen"
+            type="motor_driver.py"     
+     >
+        <param name="driver_pinout"              value="$(arg driver_L)"/>         
+        <remap from="/wheel_A_ctrl/command" to="/front_left_wheel_ctrl/command"/>
+        <remap from="/wheel_B_ctrl/command" to="/back_left_wheel_ctrl/command"/>    
+    </node>
```
Comment (cychitivav) [inline launch/run_robot.launch]: Arguments cannot be loaded with `command="load"`

```suggestion
    <rosparam file="$(find pai)/config/driver_R_pinout.yaml" command="load" />
    <node   name="right_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py" 
        >
        <param name="driver_pinout"              value="$(arg driver_R)"/>     
        <remap from="/wheel_A_ctrl/command" to="/front_right_wheel_ctrl/command"/>
        <remap from="/wheel_B_ctrl/command" to="/back_right_wheel_ctrl/command"/>
    </node>

    
    <!-- Left differential driver  -->
    <rosparam file="$(find pai)/config/driver_L_pinout.yaml" command="load" />
    <node   name="left_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py"     
     >
        <param name="driver_pinout"              value="$(arg driver_L)"/>         
        <remap from="/wheel_A_ctrl/command" to="/front_left_wheel_ctrl/command"/>
        <remap from="/wheel_B_ctrl/command" to="/back_left_wheel_ctrl/command"/>    
    </node>
```

File: scripts/motor_driver.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+try:
+    pin = rospy.get_param('driver_pinout')
+    rospy.logwarn("no 'driver_pinout' param found using default config file")
+except:
+    try:
+        (pin, namespace) =  rosparam.load_file("./config/driver_L_pinout.yaml", )
+        rospy.logwarn("no default config file using hard coded pinout")
+    except:
+        pin ={ 'FAULT':23, 'PWM2':24, 'PWM1':12, 'DIR2':16, 'DIR1':20, 'EN': 21}
```
Comment (cychitivav) [inline scripts/motor_driver.py]: This is a bit exaggerated, better to pass the default parameters and get the parameters using as namespace the node's name.
```suggestion
pin = rospy.get_param(rospy.get_name(), {'FAULT':23, 'PWM2':24, 'PWM1':12, 'DIR2':16, 'DIR1':20, 'EN': 21})
```

File: scripts/motor_driver.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        Bool_msg = Bool
+        Bool_msg.data = self.getFault()
+        rospy.logwarn("driver fault detected")
+        self.pub_fault_warning(Bool_msg)
+
+        
+
+
+if __name__ == '__main__':
+    rospy.init_node("Computer", anonymous=True)
```
Comment (cychitivav) [inline scripts/motor_driver.py]: A little suggestion
```suggestion
    rospy.init_node("Driver", anonymous=True)
```

File: scripts/motor_driver.py
Comment (cychitivav) [inline scripts/motor_driver.py:1]: As I said before, I prefer not to use the dual_max14870_rpi library.

**Review by cychitivav — CHANGES_REQUESTED** [review]

File: LICENSE
Comment (jsduenass) [reply LICENSE:1]: Alright I will fork the library repo and then added here as a submodule

File: lib/max14870-driver/LICENSE.txt
Comment (jsduenass) [reply lib/max14870-driver/LICENSE.txt:1]: This would still hold if we use the submodule

File: config/drivers_pinout.yaml
Comment (jsduenass) [reply config/drivers_pinout.yaml:1]: I was still testing and didn't know which approach was better to use a config file for both motor or separate each in its config file. I decided the latter so I remove this file

**Comment by jsduenass** [comment]
alright @cychitivav  how do you find the changes?

File: lib/max14870-driver/LICENSE.txt
Comment (cychitivav) [reply lib/max14870-driver/LICENSE.txt:1]: and why not create our own code?

**Comment by cychitivav** [comment]
In general lgtm, but I prefer to create our own code, I see that the `Driver` class inherits from `DualMotorDriver` and I think they are the same thing. This would go against the OOP paradigm.

**Comment by jsduenass** [comment]
It is true that the Driver class is not doing that much is just working as a wrapper to communicate with the ROS messages. And the library used is not that complex. However, I think that the implementation of this change would better be handle in a separate issue and merge what we have so far.

File: launch/run_robot.launch [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    
+    <!-- Left differential driver  -->
+    <rosparam file="$(find pai)/config/driver_L_pinout.yaml" command="load" />
+    <node   name="left_driver" 
+            pkg="pai" output="screen"
+            type="motor_driver.py"     
+     >
+        <param name="driver_pinout"              value="$(arg driver_L)"/>
```
Comment (cychitivav) [inline launch/run_robot.launch]: It is also necessary to remove the `args`.
```suggestion
    <rosparam file="$(find pai)/config/driver_R_pinout.yaml" command="load" />
    <node   name="right_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py" 
        >
        <remap from="/wheel_A_ctrl/command" to="/front_right_wheel_ctrl/command"/>
        <remap from="/wheel_B_ctrl/command" to="/back_right_wheel_ctrl/command"/>
    </node>

    
    <!-- Left differential driver  -->
    <rosparam file="$(find pai)/config/driver_L_pinout.yaml" command="load" />
    <node   name="left_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py"     
     >   
```

**Review by cychitivav — CHANGES_REQUESTED** [review]

**Comment by cychitivav** [comment]
> It is true that the Driver class is not doing that much is just working as a wrapper to communicate with the ROS messages. And the library used is not that complex. However, I think that the implementation of this change would better be handle in a separate issue and merge what we have so far.

Ok, I think you are right, for now let's fix the last changes and in another PR or issue we will discuss about the submodule.

**Comment by jsduenass** [comment]
> Ok, I think you are right, for now let's fix the last changes and in another PR or issue we will discuss about the submodule

Sounds good

File: launch/run_robot.launch (hunk 7 above)
Comment (jsduenass) [reply launch/run_robot.launch]: I think with this change both nodes will be left with the parameter of the last definition. in this case `driver_L_pinout.yaml`.
but let's include the change to see how it goes.

**Review by cychitivav — APPROVED** [review]

**Pull request merged**